pkg/kwok/server/debugging_logs.go

```diff
@@ -89,3 +126,424 @@ func (s *Server) getContainerLogs(request *restful.Request, response *restful.Re
 		return
 	}
 }
+
+func (s *Server) getPodLogs(podName, podNamespace, container string) (*internalversion.Log, error) {
+	l, has := slices.Find(s.config.Logs, func(l *internalversion.Logs) bool {
+		return l.Name == podName && l.Namespace == podNamespace
+	})
+	if has {
+		log, found := findLogInLogs(container, l.Spec.Logs)
+		if found {
+			return log, nil
+		}
+	} else {
+		for _, cl := range s.config.ClusterLogs {
+			if !cl.Spec.Selector.Match(podName, podNamespace) {
+				continue
+			}
+
+			log, found := findLogInLogs(container, cl.Spec.Logs)
+			if found {
+				return log, nil
+			}
+		}
+	}
+	return nil, fmt.Errorf("Failed to find pod with container name %q", container)
+}
+
+func findLogInLogs(container string, logs []internalversion.Log) (*internalversion.Log, bool) {
+	var defaultLog *internalversion.Log
+	for i, l := range logs {
+		if len(l.Containers) == 0 && defaultLog == nil {
+			defaultLog = &logs[i]
+			continue
+		}
+		if len(container) != 0 {
+			if slices.Contains(l.Containers, container) {
+				return &l, true
+			}
+		}
+	}
+	return defaultLog, defaultLog != nil
+}
+
+// logMessage is the CRI internal log type.
+type logMessage struct {
+	timestamp time.Time
+	stream    runtimeapi.LogStreamType
+	log       []byte
+}
+
+// reset resets the log to nil.
+func (l *logMessage) reset() {
+	l.timestamp = time.Time{}
+	l.stream = ""
+	l.log = nil
+}
+
+// logOptions is the CRI internal type of all log options.
+type logOptions struct {
+	tail      int64
+	bytes     int64
+	since     time.Time
+	follow    bool
+	timestamp bool
+}
+
+// newLogOptions convert the v1.PodLogOptions to CRI internal logOptions.
+func newLogOptions(apiOpts *corev1.PodLogOptions, now time.Time) *logOptions {
+	opts := &logOptions{
+		tail:      -1, // -1 by default which means read all logs.
+		bytes:     -1, // -1 by default which means read all logs.
+		follow:    apiOpts.Follow,
+		timestamp: apiOpts.Timestamps,
+	}
+	if apiOpts.TailLines != nil {
+		opts.tail = *apiOpts.TailLines
+	}
+	if apiOpts.LimitBytes != nil {
+		opts.bytes = *apiOpts.LimitBytes
+	}
+	if apiOpts.SinceSeconds != nil {
+		opts.since = now.Add(-time.Duration(*apiOpts.SinceSeconds) * time.Second)
+	}
+	if apiOpts.SinceTime != nil && apiOpts.SinceTime.After(opts.since) {
+		opts.since = apiOpts.SinceTime.Time
+	}
+	return opts
+}
+
+func readLogs(ctx context.Context, logInfo *internalversion.Log, opts *logOptions, stdout, stderr io.Writer) error {
```

Comment:
Most of this implementation comes from https://github.com/kubernetes/kubernetes/blob/0c62b122c02bff9131b6db960042150a3638d3f3/pkg/kubelet/kuberuntime/logs/logs.go#L283

I think it conforms nicely with the way the kubelet is working with logs, but I might be wrong here

Reply:
I think that's enough 😄